torch/csrc/api/include/torch/optim/sgd.h

```diff
@@ -20,42 +21,62 @@ class InputArchive;
 namespace torch {
 namespace optim {
 
-struct TORCH_API SGDOptions {
-  /* implicit */ SGDOptions(double learning_rate);
-  TORCH_ARG(double, learning_rate);
+struct TORCH_API SGDOptions : public OptimizerCloneableOptions<SGDOptions> {
+  /* implicit */ SGDOptions(double lr);
+  TORCH_ARG(double, lr);
   TORCH_ARG(double, momentum) = 0;
   TORCH_ARG(double, dampening) = 0;
   TORCH_ARG(double, weight_decay) = 0;
   TORCH_ARG(bool, nesterov) = false;
+public:
+  void serialize(torch::serialize::InputArchive& archive) override;
+  void serialize(torch::serialize::OutputArchive& archive) const override;
+  TORCH_API friend bool operator==(const SGDOptions& lhs, const SGDOptions& rhs);
+  ~SGDOptions() = default;
+};
+
+struct TORCH_API SGDParamState : public OptimizerCloneableParamState<SGDParamState> {
+  TORCH_ARG(torch::Tensor, momentum_buffer);
+
+public:
+  void serialize(torch::serialize::InputArchive& archive) override;
+  void serialize(torch::serialize::OutputArchive& archive) const override;
+  TORCH_API friend bool operator==(const SGDParamState& lhs, const SGDParamState& rhs);
+  ~SGDParamState() = default;
 };
 
 class TORCH_API SGD : public Optimizer {
  public:
-  template <typename ParameterContainer>
-  explicit SGD(ParameterContainer&& parameters, const SGDOptions& options_)
-      : Optimizer(std::forward<ParameterContainer>(parameters)),
-        options(options_) {}
+  explicit SGD(std::vector<OptimizerParamGroup> param_groups,
+      SGDOptions defaults) : Optimizer(std::move(param_groups), std::make_unique<SGDOptions>(defaults)) {
+    TORCH_CHECK(defaults.lr() >= 0, "Invalid learning rate: ", defaults.lr());
+    TORCH_CHECK(defaults.momentum() >= 0, "Invalid momentum value: ", defaults.momentum());
+    TORCH_CHECK(defaults.weight_decay() >= 0, "Invalid weight_decay value: ", defaults.weight_decay());
+    TORCH_CHECK(!defaults.nesterov() || (defaults.momentum() > 0 && defaults.dampening() == 0), "Nesterov momentum requires a momentum and zero dampening");
+    for (const auto& group : param_groups_) {
```

Comment:
python version for sgd doesn't initialize the state here but it does for adagrad. so, even though we are aiming to achieve parity between python and C++ API, I thought it would be nice to have some consistency within each API. what do you think?

Reply:
also should we have a state init function to add all params to state when a new param_group is added?

Reply:
if we don't initialize the state here, the test fails here: https://github.com/pytorch/pytorch/blob/fe3eb09da5783889938ff71bb548b7277c5fa607/test/cpp/api/serialize.cpp#L131 because state size for the optimizer3 is 0 
so I think we should definitely either add checks in step to ensure new params are added to the state if they don't already exist or have state_init